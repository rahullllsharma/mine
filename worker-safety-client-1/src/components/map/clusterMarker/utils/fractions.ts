type RiskPointProperties = {
  risk: string;
  className: string;
  start: number;
  end: number;
};

function calculateFractionsPerRisk(
  totalsAsAssoc: [string, number | undefined][],
  total: number
): RiskPointProperties[] {
  let lastStartPoint = 0;

  const totalsPerFraction = totalsAsAssoc
    // exclude risks without scores
    .filter(([, value]) => !!value)
    .map(([key, value]) => {
      // determine the start and end points.
      const start = lastStartPoint;

      const segment = parseFloat(((value as number) / total).toFixed(3));
      const end = start + Math.abs(segment * 360);

      // replace the last point by the new one.
      lastStartPoint = end;

      return {
        risk: key,
        className: `text-risk-${key}`,
        start,
        end: end < 360 ? end : 359.9,
      };
    });

  return totalsPerFraction;
}

export { calculateFractionsPerRisk };
