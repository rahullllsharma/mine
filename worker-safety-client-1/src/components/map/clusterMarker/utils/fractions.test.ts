import { calculateFractionsPerRisk } from "./fractions";

describe(calculateFractionsPerRisk.name, () => {
  it.each`
    high  | medium       | low          | unknown
    ${25} | ${25}        | ${25}        | ${25}
    ${25} | ${25}        | ${25}        | ${undefined}
    ${25} | ${25}        | ${undefined} | ${undefined}
    ${25} | ${undefined} | ${undefined} | ${undefined}
    ${25} | ${25}        | ${25}        | ${25}
    ${25} | ${25}        | ${25}        | ${0}
    ${25} | ${25}        | ${0}         | ${0}
    ${25} | ${0}         | ${0}         | ${0}
  `(
    "should calculate the proper paths and remove risks without scores",
    scores => {
      const params = Object.entries(scores) as [string, number | undefined][];
      const total = params.reduce((acc, [, val]) => acc + (val ?? 0), 0);

      const result = calculateFractionsPerRisk(params, total);
      expect(result).toMatchSnapshot();

      expect(result.length).toEqual(
        params.filter(([, value]) => !!value).length
      );

      // Remove all the undefined elements
      params.forEach(([key, value]) => {
        if (!value) {
          expect(
            result.some(fraction => fraction.className.includes(key))
          ).toBe(false);
        }
      });
    }
  );
});
