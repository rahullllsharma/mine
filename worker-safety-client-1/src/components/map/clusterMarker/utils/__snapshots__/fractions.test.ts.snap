// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 1`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 90,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 180,
    "risk": "medium",
    "start": 90,
  },
  Object {
    "className": "text-risk-low",
    "end": 270,
    "risk": "low",
    "start": 180,
  },
  Object {
    "className": "text-risk-unknown",
    "end": 359.9,
    "risk": "unknown",
    "start": 270,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 2`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 119.88000000000001,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 239.76000000000002,
    "risk": "medium",
    "start": 119.88000000000001,
  },
  Object {
    "className": "text-risk-low",
    "end": 359.64000000000004,
    "risk": "low",
    "start": 239.76000000000002,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 3`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 180,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 359.9,
    "risk": "medium",
    "start": 180,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 4`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 359.9,
    "risk": "high",
    "start": 0,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 5`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 90,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 180,
    "risk": "medium",
    "start": 90,
  },
  Object {
    "className": "text-risk-low",
    "end": 270,
    "risk": "low",
    "start": 180,
  },
  Object {
    "className": "text-risk-unknown",
    "end": 359.9,
    "risk": "unknown",
    "start": 270,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 6`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 119.88000000000001,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 239.76000000000002,
    "risk": "medium",
    "start": 119.88000000000001,
  },
  Object {
    "className": "text-risk-low",
    "end": 359.64000000000004,
    "risk": "low",
    "start": 239.76000000000002,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 7`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 180,
    "risk": "high",
    "start": 0,
  },
  Object {
    "className": "text-risk-medium",
    "end": 359.9,
    "risk": "medium",
    "start": 180,
  },
]
`;

exports[`calculateFractionsPerRisk should calculate the proper paths and remove risks without scores 8`] = `
Array [
  Object {
    "className": "text-risk-high",
    "end": 359.9,
    "risk": "high",
    "start": 0,
  },
]
`;
